<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Привет еще раз. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<b>Комментарий от ревьюера</b>

Привет! Меня зовут Ярослав Кальманов, я буду делать ревью твоего проекта. Давай будем общаться на «ты».

Мои комментарии будут в ячейках "Комментарий от ревьюера".

<div class="alert alert-danger"> 
Красным цветом обозначены комментарии, на которые тебе нужно обратить внимание в первую очередь, чтобы я принял проект. 
</div>
<div class="alert alert-warning">
Желтым цветом отмечено то, что ты можешь сделать немного по-другому. Ты можешь учесть эти комментарии при выполнении следующих проектов.
</div>
<div class="alert alert-success">
Зелёным цветом обозначены твои удачные решения, которые ты без сомнения можешь применять в будущих проектах.  
</div>

Если ты вносишь изменения в проект по моим рекомендациям, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь свои комментарии заметным цветом или поместишь 

<div class="alert alert-info">
    
в такую ячейку
</div>

Давай при работе с проектом мы будем общаться с тобой в диалоге, поэтому не удаляй и не перемещай мои комментарии, это поможет мне сделать повторную  проверку проекта быстрее.
___

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


**Цель исследования**:
Зависит ли семейное положение и количество детей клиента на факт погашения кредита в срок.

**Ход исследования**
 1. Обзор данных
 2. Предобработка данных
 3. Ответы на вопросы
 4. Общий вывод
 5. Рекомендации банку

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Хорошее начало
</div>



## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
# импорт библиотеки pandas
import pandas as pd

In [2]:
# чтение файла с данными и сохранение в data
data = pd.read_csv('/datasets/data.csv')

In [3]:
# получение первых 10 строк таблицы data
display(data.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
# получение общей информации о данных в таблице data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
# обзор значений столбцов с типами данных int64 и float64
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

В таблице двенадцать столбцов. В столбцах встречаются три типа данных:
* `int64` (5 столбцов);
* `float64` (2 столбца);
* `object` (5 столбцов).

Согласно документации к данным:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита.

В названиях колонок нет нарушений стиля, но некоторые наименования неточно отражают смысл хранящихся данных.
Поэтому их стоит переименовать.

Некоторые типы данных лучше сменить для корректности анализа.

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


**Вывод**

В каждой строке таблицы — данные клиентов банка. Для проверки гипотезы о прямой зависимости числа детей и семейного положения на факт погашения кредита данных более чем достаточно.
Но в данных много недочетов и неточностей. Ниже перечислены проблемы, найденные в таблице:

- неточные названия столбцов;
- минимальное значение в столбце о возрасте клиентов - 0;
- наличие отрицательных значений в столбце с количеством детей;
- максимальное количество детей - 20;
- наличие отрицательных значений в столбце с количеством дней трудового стажа;
- пропуски в столбцах о ежемесячном доходе и днях трудового стажа;
- в некоторых столбцах тип данных не совпадает со смыслом описываемой переменной;
- в некоторых столбцах лучше конвертировать величины (столбец о днях трудового стажа);
- в столбце об образовании наименования указаны в разном регистре;
- есть избыточные данные — текстовое описание категории, а также присвоенное ей число (столбцы о семейном положении и образовании).

Перед проверкой гипотезы необходимо устранить указанные проблемы в данных.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все верно
</div>



## Шаг 2. Предобработка данных

**Этапы выполнения**:
 1. Переименовать столбцы;
 2. Исправить некорректные значения в столбцах;
 2. Определить и заполнить пропущенные значения:
  * описать, какие пропущенные значения обнаружены;
  * привести возможные причины появления пропусков в данных;
  * объяснить, по какому принципу заполнены пропуски;
 3. Заменить вещественный тип данных на целочисленный:
  * пояснить, как выбирался метод для изменения типа данных; 
 4. Удалить дубликаты:
  * пояснить, как выбирался метод для поиска и удаления дубликатов в данных;
  * привести возможные причины появления дубликатов;
 5. Выделить леммы в значениях столбца с целями получения кредита:
  * описать, как проводилась лемматизация целей кредита;
  * категоризировать данные:
  * перечислить, какие «словари» выделены для этого набора данных, и объяснить, почему.

### Изменение некорректных имен столбцов

In [6]:
# переименовывание названий столбцов методом rename()
data = data.rename(columns = {'children': 'children_number',
                             'dob_years': 'age_in_years',
                              'income_type': 'employment_type',
                              'total_income': 'monthly_income',
                             'purpose': 'credit_purpose'})

In [7]:
# проверка того что столбцы корректно переменовались
display(data.head(5))

children_number  days_employed  age_in_years education  education_id  \
0                1   -8437.673028            42    высшее             0   
1                1   -4024.803754            36   среднее             1   
2                0   -5623.422610            33   Среднее             1   
3                3   -4124.747207            32   среднее             1   
4                0  340266.072047            53   среднее             1   

      family_status  family_status_id gender employment_type  debt  \
0   женат / замужем                 0      F       сотрудник     0   
1   женат / замужем                 0      F       сотрудник     0   
2   женат / замужем                 0      M       сотрудник     0   
3   женат / замужем                 0      M       сотрудник     0   
4  гражданский брак                 1      F       пенсионер     0   

   monthly_income              credit_purpose  
0   253875.639453               покупка жилья  
1   112080.014102     приобретение автомобиля  
2   145885.952297               покупка жилья  
3   267628.550329  дополнительное образование  
4   158616.077870             сыграть свадьбу

<div class="alert alert-warning">
    
<b> Комментарий от ревьюера</b>
    
В последней строке ячейки можно не использовать функции print()/display(). Результат работы кода и без них будет выведен на экран
</div>

<div class="alert alert-info">

<b>Комментарий студента</b>
    
Привет, Ярослав! Спасибо за твои комментарии :)
Иногда путаюсь и добавляю лишний display().
</div>

**Вывод**

Стобцы переименованы в более удобный для восприятия вид

### Точечное изменение некорректных значений

**Исправление значений 0 в столбце `age_in_years`**

In [8]:
# нахождение минимального возраста и следующего за ним значения
print("Минимальный возраст клиента:", sorted(data['age_in_years'].unique())[0], "лет")
print("Следующее за ним значение возраста:" , sorted(data['age_in_years'].unique())[1], "лет")

Минимальный возраст клиента: 0 лет
Следующее за ним значение возраста: 19 лет


Следующее минимальное значение возраста клиента - 19, а выдача кредита банком возможна с 18 лет.

In [9]:
# замена значения 0 в столбце `age_in_years` на 18
data.loc[data['age_in_years'] == 0, 'age_in_years'] = 18

<div class="alert alert-warning">
    
<b> Комментарий от ревьюера</b>
    
Думаю, не стоит заменять 0 на 18. Например, у таких клиентов может быть employment_type "Пенсионер" или вроде того, и наша замена будет некорректной.
    
Если мы не можем точно восстановить данные, лучше их не изменять
</div>

<b>Комментарий студента</b>
<div class="alert alert-info">
Поняла. Или уточнять что за данные у заказчика.
</div>

<div class="alert alert-success">
    
<b> Комментарий от ревьюера 2</b>

Ты права. Если есть возможность уточнить у коллег, лучше это сделать
</div>

In [10]:
# проверка правильности замены
print("Минимальный возраст клиента: ", data['age_in_years'].min(), "лет")

Минимальный возраст клиента:  18 лет


**Исправление отрицательных значений в столбце `children_number`**

In [11]:
data[data['children_number'] < 0]['children_number'].unique()

array([-1])

В столбце о количестве детей есть отрицательное значение "-1". 
Возможно, это техническая ошибка, поэтому оно заменяется на "1".

In [12]:
data.loc[data['children_number'] == -1, 'children_number'] = 1

In [13]:
# проверка правильности замены
print("Минимальное количество детей: ", data['children_number'].min())

Минимальное количество детей:  0


Также стоит проверить, какие вообще уникальные положительные значения есть

In [14]:
# применение метода unique
data['children_number'].unique()

array([ 1,  0,  3,  2,  4, 20,  5])

Промежуточных значений между 5 и 20 нет. Можно посчитать ошибкой. Но необходимо проверить, сколько всего строк со значением 20 в столбце `children_number`.

In [15]:
# проверка числа строк
data[data['children_number'] == 20]

children_number  days_employed  age_in_years education  education_id  \
606                 20    -880.221113            21   среднее             1   
720                 20    -855.595512            44   среднее             1   
1074                20   -3310.411598            56   среднее             1   
2510                20   -2714.161249            59    высшее             0   
2941                20   -2161.591519            18   среднее             1   
...                ...            ...           ...       ...           ...   
21008               20   -1240.257910            40   среднее             1   
21325               20    -601.174883            37   среднее             1   
21390               20            NaN            53   среднее             1   
21404               20    -494.788448            52   среднее             1   
21491               20    -173.954460            27   среднее             1   

         family_status  family_status_id gender employment_type  debt  \
606    женат / замужем                 0      M       компаньон     0   
720    женат / замужем                 0      F       компаньон     0   
1074   женат / замужем                 0      F       сотрудник     1   
2510    вдовец / вдова                 2      F       сотрудник     0   
2941   женат / замужем                 0      F       сотрудник     0   
...                ...               ...    ...             ...   ...   
21008  женат / замужем                 0      F       сотрудник     1   
21325  женат / замужем                 0      F       компаньон     0   
21390  женат / замужем                 0      M       компаньон     0   
21404  женат / замужем                 0      M       компаньон     0   
21491  женат / замужем                 0      F       сотрудник     0   

       monthly_income                         credit_purpose  
606     145334.865002                          покупка жилья  
720     112998.738649                   покупка недвижимости  
1074    229518.537004                  получение образования  
2510    264474.835577  операции с коммерческой недвижимостью  
2941    199739.941398                  на покупку автомобиля  
...               ...                                    ...  
21008   133524.010303                        свой автомобиль  
21325   102986.065978                 профильное образование  
21390             NaN             покупка жилой недвижимости  
21404   156629.683642        операции со своей недвижимостью  
21491   155520.899688     на покупку подержанного автомобиля  

[76 rows x 12 columns]

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Посчитать количество повторений значения можно так:  
`data.children_number.value_counts()`
</div>


<div class="alert alert-info">

<b>Комментарий студента</b>

Ок
</div>

Таких значений только 76 ( из почти 22000 строк таблицы). Поэтому можно заменить значение 20 на 2.

In [16]:
# замена количества детей с 20 на 2 с помощью loc()
data.loc[data['children_number'] == 20, 'children_number'] = 2

In [17]:
# проверка сделанных изменений
data['children_number'].unique()

array([1, 0, 3, 2, 4, 5])

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Здорово, что ты нашла и исправила ошибки в children_number
</div>

**Исправление значений в столбце `days_employed`**

In [18]:
# обзор значений столбца `days_employed`
data['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Если предположить, что человек работал с 0 до 100 лет, что очень маловероятно, то его трудовой стаж в днях составляет примерно 36500 дней.
Необходимо проверить, какие значения записаны отрицательными и положительными числами.

In [19]:
# проверка положительных значений data['days_employed'] по заданным условиям
data[(data['days_employed'] >= 0) & (data['days_employed'] < 36500)]

Empty DataFrame
Columns: [children_number, days_employed, age_in_years, education, education_id, family_status, family_status_id, gender, employment_type, debt, monthly_income, credit_purpose]
Index: []

Таких значений нет. Далее надо проверить, как обстоят дела с отрицательными значениями трудового стажа.

In [20]:
# проверка отрицательных значений data['days_employed'] по заданным условиям
data[(data['days_employed'] <= 0) & (data['days_employed'] < 36500)]

children_number  days_employed  age_in_years       education  \
0                    1   -8437.673028            42          высшее   
1                    1   -4024.803754            36         среднее   
2                    0   -5623.422610            33         Среднее   
3                    3   -4124.747207            32         среднее   
5                    0    -926.185831            27          высшее   
...                ...            ...           ...             ...   
21519                1   -2351.431934            37  ученая степень   
21520                1   -4529.316663            43         среднее   
21522                1   -2113.346888            38         среднее   
21523                3   -3112.481705            38         среднее   
21524                2   -1984.507589            40         среднее   

       education_id     family_status  family_status_id gender  \
0                 0   женат / замужем                 0      F   
1                 1   женат / замужем                 0      F   
2                 1   женат / замужем                 0      M   
3                 1   женат / замужем                 0      M   
5                 0  гражданский брак                 1      M   
...             ...               ...               ...    ...   
21519             4         в разводе                 3      M   
21520             1  гражданский брак                 1      F   
21522             1  гражданский брак                 1      M   
21523             1   женат / замужем                 0      M   
21524             1   женат / замужем                 0      F   

      employment_type  debt  monthly_income                     credit_purpose  
0           сотрудник     0   253875.639453                      покупка жилья  
1           сотрудник     0   112080.014102            приобретение автомобиля  
2           сотрудник     0   145885.952297                      покупка жилья  
3           сотрудник     0   267628.550329         дополнительное образование  
5           компаньон     0   255763.565419                      покупка жилья  
...               ...   ...             ...                                ...  
21519       сотрудник     0   115949.039788  покупка коммерческой недвижимости  
21520       компаньон     0   224791.862382                  операции с жильем  
21522       сотрудник     1    89672.561153                       недвижимость  
21523       сотрудник     1   244093.050500       на покупку своего автомобиля  
21524       сотрудник     0    82047.418899              на покупку автомобиля  

[15906 rows x 12 columns]

Такие значения дней трудового стажа уже могут быть реальными. 
Возможно, выгрузка происходила из разных кредитных систем, где значения записываются в других форматах, возможно, не в днях.
Помимо этого, возможен человеческий фактор (данные заносили разные люди и кто-то из них ошибся).
Отрицательные-скорее всего, в днях, как и указано в описании к столбцу. А положительные-пока неясно.

Но в силу того что для проверки гипотезы необходимо учитывать семейное положение и количество детей,то для удобства восприятия надо сделать все отрицательные значения положительными, а дни перевести в года, создав дополнительный столбец в таблице.
Для начала надо сделать все числа положительными путем нахождения их модуля.

In [21]:
# взятие модуля числа
data['days_employed'] = data['days_employed'].abs()

In [22]:
# проверка значений столбца `days_employed`
data['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [23]:
# перевод дней трудового стажа в года и добавление нового столбца
data['years_employed'] = data['days_employed'] / 365

In [24]:
# удаление столбца `days_employed`
del data['days_employed']

In [25]:
data.head(10)

children_number  age_in_years education  education_id     family_status  \
0                1            42    высшее             0   женат / замужем   
1                1            36   среднее             1   женат / замужем   
2                0            33   Среднее             1   женат / замужем   
3                3            32   среднее             1   женат / замужем   
4                0            53   среднее             1  гражданский брак   
5                0            27    высшее             0  гражданский брак   
6                0            43    высшее             0   женат / замужем   
7                0            50   СРЕДНЕЕ             1   женат / замужем   
8                2            35    ВЫСШЕЕ             0  гражданский брак   
9                0            41   среднее             1   женат / замужем   

   family_status_id gender employment_type  debt  monthly_income  \
0                 0      F       сотрудник     0   253875.639453   
1                 0      F       сотрудник     0   112080.014102   
2                 0      M       сотрудник     0   145885.952297   
3                 0      M       сотрудник     0   267628.550329   
4                 1      F       пенсионер     0   158616.077870   
5                 1      M       компаньон     0   255763.565419   
6                 0      F       компаньон     0   240525.971920   
7                 0      M       сотрудник     0   135823.934197   
8                 1      F       сотрудник     0    95856.832424   
9                 0      M       сотрудник     0   144425.938277   

               credit_purpose  years_employed  
0               покупка жилья       23.116912  
1     приобретение автомобиля       11.026860  
2               покупка жилья       15.406637  
3  дополнительное образование       11.300677  
4             сыграть свадьбу      932.235814  
5               покупка жилья        2.537495  
6           операции с жильем        7.888225  
7                 образование        0.418574  
8       на проведение свадьбы       18.985932  
9     покупка жилья для семьи        5.996593

В столбце `years_employed` остались нереальные значения трудового стажа (932 с лишним года).
Придется оставить строки, содержащие эти столбцы, для лемматизации и для сохранения полноты анализа. 
А также далее необходимо поменять тип столбца `years_employed` с float64 на int64 чтобы избавиться от дробной части.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все хорошо. 
    
Может быть у клиентов со стажем в 932 года признак days_employed был указан в часах, а не днях или десятичный разделитель смещен вправо на пару символов.
</div>


<div class="alert alert-info">

<b>Комментарий студента</b>

Да, значения этого столбца для меня были самые непонятные
</div>

**Проверка уникальных значений столбца `gender`**

In [26]:
# проверка с помощью метода unique()
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [27]:
# количество строк, в которых gender - `XNA`
data[data['gender'] == 'XNA']

children_number  age_in_years            education  education_id  \
10701                0            24  неоконченное высшее             2   

          family_status  family_status_id gender employment_type  debt  \
10701  гражданский брак                 1    XNA       компаньон     0   

       monthly_income        credit_purpose  years_employed  
10701   203905.157261  покупка недвижимости        6.461919

Пол "XNA" встречается только в одной строке таблицы. Всего строк-21525.
Удаление одной строки незначительно повлияет на результат проверки гипотезы, поэтому ее можно удалить.

In [28]:
# удаление строки
data = data.drop(data[data['gender'] == 'XNA'].index)

In [29]:
# проверка
print(data['gender'].unique())
data[data['gender'] == 'XNA']

['F' 'M']


Empty DataFrame
Columns: [children_number, age_in_years, education, education_id, family_status, family_status_id, gender, employment_type, debt, monthly_income, credit_purpose, years_employed]
Index: []

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Молодец, что изучила все признаки
</div>


### Обработка пропусков

In [30]:
# подсчет числа пропусков в каждой колонке
data.isna().sum()

children_number        0
age_in_years           0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
employment_type        0
debt                   0
monthly_income      2174
credit_purpose         0
years_employed      2174
dtype: int64

Пропущенные значения есть в колонках `years_employed` (2174 раз) и `monthly_income` (2174 раз) (общий трудовой стаж в днях и ежемесячный доход клиента соответственно). 
Колонки с этими данными нельзя исключить из таблицы, так как они являются одими из важных критериев одобрения кредита.
Пропуски в этих столбцах появились, возможно, потому что человеку выплачивалась "серая" зарплата или сведений об этих данных нет. Также вероятен человеческий фактор.

Проверим, что значения в ячейках этих столбцов пропущены одновременно.

In [31]:
# подсчет одновременно пропущенных значений есть в колонках `days_employed` и `monthly_income`
len(data[data['years_employed'].isna() & data['monthly_income'].isna()])

2174

In [32]:
# визуальная оценка
display(data[data['years_employed'].isna() & data['monthly_income'].isna()].head())

children_number  age_in_years education  education_id  \
12                0            65   среднее             1   
26                0            41   среднее             1   
29                0            63   среднее             1   
41                0            50   среднее             1   
55                0            54   среднее             1   

            family_status  family_status_id gender employment_type  debt  \
12       гражданский брак                 1      M       пенсионер     0   
26        женат / замужем                 0      M     госслужащий     0   
29  Не женат / не замужем                 4      F       пенсионер     0   
41        женат / замужем                 0      F     госслужащий     0   
55       гражданский брак                 1      F       пенсионер     1   

    monthly_income                    credit_purpose  years_employed  
12             NaN                   сыграть свадьбу             NaN  
26             NaN                       образование             NaN  
29             NaN  строительство жилой недвижимости             NaN  
41             NaN  сделка с подержанным автомобилем             NaN  
55             NaN                   сыграть свадьбу             NaN

Значения в ячейках этих столбцов пропущены одновременно. Удалять строки с пропущенными значениями нельзя. Значит можно провести группировку по столбцам `age_in_years`и `gender` , а затем для каждой группы найти медиану и присвоить ее пропускам из той же группы


**Сначала производится заполнение пропусков в столбце `years_employed`**

In [33]:
# построение сводной таблицы по полу и возрасту
# далее для каждой группы считается медианное значение трудового стажа
days_employed_pivot = data.pivot_table(index=['age_in_years'],
                                columns = 'gender',
                                values = 'years_employed',
                                aggfunc='median')

In [34]:
# первые 10 строк сводной таблицы
days_employed_pivot.head(10)

gender               F         M
age_in_years                    
18            6.188825  2.840121
19            1.945838  2.425393
20            1.631669  2.412408
21            1.665763  1.718317
22            1.912607  2.029730
23            1.738662  1.993697
24            2.643547  2.469998
25            2.605977  2.380990
26            3.237984  2.640139
27            3.465408  3.075200

In [35]:
# определение функции по заполнению пропусков в столбце 'days_employed' медианными значениями
# с учетом группировки по полу и возрасту
def days_employed_fill(data):
    """Функция проверяет значения в столбце 'days_employed'.
    Если оно пропущено, то меняет на медианное значение трудового стажа
    с учетом принадлежности к такой же группе по полу и возрасту.
    Иначе, оставляет прежнее значение.
    """
    gender = data['gender']
    age = data['age_in_years']
    if pd.isna(data['years_employed']):
        return days_employed_pivot.loc[age,gender]
    else:
        return data['years_employed']

In [36]:
# заполнение пропусков в столбце 'years_employed' с использованием функции days_employed_fill
data['years_employed'] = data.apply(days_employed_fill, axis = 1)

In [37]:
# подсчет числа пропусков в колонке 'years_employed'
data['years_employed'].isna().sum()

0

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Посмотри [тут](https://stackoverflow.com/questions/19966018/pandas-filling-missing-values-by-mean-in-each-group), как можно заполнить пропуски групповыми медианами без создания дополнительных переменных и функций
</div>

<div class="alert alert-info">

<b>Комментарий студента</b>
    
Спасибо :)
</div>

**Осталось заполнить пропуски в столбце `monthly_income`**

In [38]:
# обзор значений столбца 'monthly_income' с округлением до целых
data['monthly_income'].describe().astype(int)

count      19350
mean      167420
std       102973
min        20667
25%       103040
50%       145011
75%       203424
max      2265604
Name: monthly_income, dtype: int64

Можно заполнить пропуски медианным значениями общего месячного дохода, но так не следует делать из-за большого разброса значений.
Помимо этого максимальный месячный доход почти в 110 раз превышает минимальный, а также сумма в 2 миллиона очень велика для ежемесячного дохода. Следовательно стоит заполнить пропуски медианными значениями ежемесячного дохода опираясь на пол и возраст.

In [39]:
monthly_income_pivot = data.pivot_table(index=['age_in_years'],
                                    columns = 'gender',
                                    values = 'monthly_income',
                                    aggfunc='median')

In [40]:
monthly_income_pivot.head(10)

gender                    F              M
age_in_years                              
18            117279.001914  179138.212343
19            112217.756662   91876.754772
20            101512.989536  134750.749515
21            112778.853259  147812.788085
22            111633.034748  158258.854591
23            117755.877467  137544.158776
24            122193.154166  153115.076039
25            131764.441672  161883.762815
26            132408.440852  165998.158511
27            142601.204718  170124.764755

In [41]:
# определение функции по заполнению пропусков в столбце 'monthly_income' медианными значениями
# с учетом группировки по полу и возрасту
def monthly_income_fill(data):
    """Функция проверяет значения в столбце 'monthly_income'.
    Если оно пропущено, то меняет на медианное значение ежемесячного дохода
    с учетом принадлежности к такой же группе по полу и возрасту.
    Иначе, оставляет прежнее значение.
    """
    gender = data['gender']
    age = data['age_in_years']
    if pd.isna(data['monthly_income']):
        return monthly_income_pivot.loc[age,gender]
    else:
        return data['monthly_income']

In [42]:
# заполнение пропусков в столбце 'monthly_income' с использованием функции monthly_income_fill
data['monthly_income'] = data.apply(monthly_income_fill, axis = 1)

In [43]:
# подсчет числа пропусков в колонке 'monthly_income'
data['monthly_income'].isna().sum()

0

In [44]:
# обзор данных полученной таблицы
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
children_number     21524 non-null int64
age_in_years        21524 non-null int64
education           21524 non-null object
education_id        21524 non-null int64
family_status       21524 non-null object
family_status_id    21524 non-null int64
gender              21524 non-null object
employment_type     21524 non-null object
debt                21524 non-null int64
monthly_income      21524 non-null float64
credit_purpose      21524 non-null object
years_employed      21524 non-null float64
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


In [45]:
# проверка отсутствия пропусков в таблице
print(data.isna().sum())

children_number     0
age_in_years        0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
employment_type     0
debt                0
monthly_income      0
credit_purpose      0
years_employed      0
dtype: int64


Пропусков в таблице нет.

**Вывод**

Пропуски в таблице заполнены на основе зависящих признаков-возраста и пола.
Пропуски в этих столбцах появились, возможно, потому что человеку выплачивалась "серая" зарплата или сведений об этих данных нет. Также вероятен человеческий фактор.
В дальнейшем необходимо спрашивать тех, кто предоставил данные для анализа о возможных причинах пропуска значений в столбцах.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Ты хорошо справилась с обработкой пропусков
</div>

### Замена типа данных

In [46]:
# обзор данных полученной таблицы
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
children_number     21524 non-null int64
age_in_years        21524 non-null int64
education           21524 non-null object
education_id        21524 non-null int64
family_status       21524 non-null object
family_status_id    21524 non-null int64
gender              21524 non-null object
employment_type     21524 non-null object
debt                21524 non-null int64
monthly_income      21524 non-null float64
credit_purpose      21524 non-null object
years_employed      21524 non-null float64
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


Значения в столбцах `monthly_income` и `years_employed` записаны в формате `float64`, его лучше сменить на `int64` для того чтобы избавиться от наличия знаков после запятой в расчетах.
В результатах понадобятся только целые числа.

In [47]:
# смена формата данных в столбце 'years_employed' с float64 на int64
data['years_employed'] = data['years_employed'].astype(int)

In [48]:
# смена формата данных в столбце 'monthly_income' с float64 на int64
data['monthly_income'] = data['monthly_income'].astype(int)

In [49]:
# проверка данных полученной таблицы
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
children_number     21524 non-null int64
age_in_years        21524 non-null int64
education           21524 non-null object
education_id        21524 non-null int64
family_status       21524 non-null object
family_status_id    21524 non-null int64
gender              21524 non-null object
employment_type     21524 non-null object
debt                21524 non-null int64
monthly_income      21524 non-null int64
credit_purpose      21524 non-null object
years_employed      21524 non-null int64
dtypes: int64(7), object(5)
memory usage: 2.1+ MB


**Вывод**

При изменении типа данных использовался метод astype(), так как именно он переводит числа, хранящиеся в pandas.Series в целочисленный формат int64.
Теперь все числа в таблице представлены в удобном формате для последующих расчетов.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все верно
</div>

### Обработка дубликатов

Все строки исследуемой таблицы должны быть очищены от явных и неявных дубликатов.

**Проверка на явные дубликаты**

In [50]:
# сначала явные дубликаты находятся методом duplicated() для подсчета явных дубликатов
# затем их количество подсчитывается методом sum()
data.duplicated().sum()

54

In [51]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
data = data.drop_duplicates().reset_index(drop = True)

In [52]:
# проверка на отсутствие дубликатов
data.duplicated().sum()

0

**Проверка на неявные дубликаты**

In [53]:
# проверка столбца 'education' на наличие дубликатов методом unique()
data["education"].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

В столбце дубликаты отличаются по регистру, это можно легко исправить методом lower()

In [54]:
# исправление дубликатов методом lower()
data["education"] = data["education"].str.lower()

In [55]:
# повторная проверка
data["education"].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [56]:
# проверка столбца 'family_status' на наличие дубликатов методом unique()
data["family_status"].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Аналогично, исправление методом lower()

In [57]:
# исправление дубликатов методом lower()
data["family_status"] = data["family_status"].str.lower()

In [58]:
# повторная проверка
data["family_status"].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

In [59]:
# проверка столбца 'employment_type' на наличие дубликатов методом unique()
data["employment_type"].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

В этом столбце обработка дубликатов не требуется

In [60]:
# повторная проверка на явные дубликаты
data.duplicated().sum()

17

In [61]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
data = data.drop_duplicates().reset_index(drop = True)

In [62]:
# финальная повторная проверка на явные дубликаты
data.duplicated().sum()

0

**Вывод**

Данные таблицы полностью очищены от явных и неявных дубликатов.
Дубликаты различались только по регистру.
Возможная причина появления: данные вносили разные люди. Кто-то начинал писать с заглавной буквы, кто-то с прописной. А кто-то и вовсе не обращал внимания на то что зажата клавиша Caps Lock.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Здорово, что ты обработала явные и неявные дубликаты
</div>

### Лемматизация

Для начала необходимо импортировать библиотеку с функцией лемматизации на русском языке и специальный контейнер "Counter" из модуля "collections" для подсчета числа лемматизированных слов.

<div class="alert alert-warning">
    
<b> Комментарий от ревьюера</b>
    
Все библиотеки лучше импортировать в начале файла. Об этом говорится в [стилевом гайде по питону](https://www.python.org/dev/peps/pep-0008/#imports)
    
    Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
</div>

<div class="alert alert-info">

<b>Комментарий студента</b>

Хорошо, поняла
</div>

In [63]:
# импорт необходимых библиотек
from pymystem3 import Mystem
from collections import Counter

In [64]:
# присвоение переменной m метода Mystem()
m = Mystem()

In [65]:
# количество уникальных целей
unique_credit_purposes = data['credit_purpose'].unique()
unique_credit_purposes

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

После изучения различных причин брать кредит, можно выделить несколько основных категорий:
- жилье;
- автомобиль;
- образование;
- свадьба;
- недвижимость;
- строительство;
- ремонт.

In [66]:
# список типов целей брать кредит
list_purposes = ['жилье', 'автомобиль', 'образование', 'свадьба',
                 'недвижимость', 'строительство', 'ремонт']

In [67]:
# определение функции по извлечению лемм и сопоставление их с основными целями кредита
def find_purpose_type(credit_purpose):
    """Функция проверяет, есть ли лемма из list_purposes в ячейке credit_purpose.
    Если есть, возвращает ее.
    """
    for lemma in m.lemmatize(credit_purpose):
        if lemma in list_purposes:
            return lemma

In [68]:
# применение функции find_purpose_type к столбцу "purpose_type"
data["purpose_type"] = data["credit_purpose"].apply(find_purpose_type)

In [69]:
# проверка правильности добавления нового столбца
data.head(10)

children_number  age_in_years education  education_id     family_status  \
0                1            42    высшее             0   женат / замужем   
1                1            36   среднее             1   женат / замужем   
2                0            33   среднее             1   женат / замужем   
3                3            32   среднее             1   женат / замужем   
4                0            53   среднее             1  гражданский брак   
5                0            27    высшее             0  гражданский брак   
6                0            43    высшее             0   женат / замужем   
7                0            50   среднее             1   женат / замужем   
8                2            35    высшее             0  гражданский брак   
9                0            41   среднее             1   женат / замужем   

   family_status_id gender employment_type  debt  monthly_income  \
0                 0      F       сотрудник     0          253875   
1                 0      F       сотрудник     0          112080   
2                 0      M       сотрудник     0          145885   
3                 0      M       сотрудник     0          267628   
4                 1      F       пенсионер     0          158616   
5                 1      M       компаньон     0          255763   
6                 0      F       компаньон     0          240525   
7                 0      M       сотрудник     0          135823   
8                 1      F       сотрудник     0           95856   
9                 0      M       сотрудник     0          144425   

               credit_purpose  years_employed purpose_type  
0               покупка жилья              23        жилье  
1     приобретение автомобиля              11   автомобиль  
2               покупка жилья              15        жилье  
3  дополнительное образование              11  образование  
4             сыграть свадьбу             932      свадьба  
5               покупка жилья               2        жилье  
6           операции с жильем               7        жилье  
7                 образование               0  образование  
8       на проведение свадьбы              18      свадьба  
9     покупка жилья для семьи               5        жилье

Из-за того что в столбце 'credit_purpose' может быть несколько лемм из списка list_purposes, необходимо повторно проверить таблицу на наличие явных дубликатов.

In [70]:
data.duplicated().sum()

0

**Вывод**

В столбце `credit_purpose` находились дублирующиеся цели кредита, которые отличались формулировкой, поэтому их понадобилось лемматизировать. Ключевые леммы записаны в отдельный столбец `purpose_type`.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все правильно
</div>

### Категоризация данных

Необходимо категоризовать данные для последующей проверки гипотез.

**Категоризация по количеству детей**

В России многодетной считается семья с 3 и более детьми, поэтому семьи клиентов можно разделить на 3 группы:
1. семья без детей;
2. семья с 1 или 2 детьми;
3. многодетная семья.

In [71]:
# определение функции разделения на 3 группы в зависимости от числа детей
def children_number_categorizer(children_number):
    """Функция разбивает семьи клиентов на 3 группы.
    """
    if children_number == 0:
        return 'семья без детей'
    elif children_number == 1 or children_number == 2:
        return 'семья с 1 или 2 детьми'
    else:
        return 'многодетная семья'

In [72]:
# применение функции
data['children_number_type'] = data['children_number'].apply(children_number_categorizer)

In [73]:
# проверка добавления столбца 'children_number_type'
data.head(5)

children_number  age_in_years education  education_id     family_status  \
0                1            42    высшее             0   женат / замужем   
1                1            36   среднее             1   женат / замужем   
2                0            33   среднее             1   женат / замужем   
3                3            32   среднее             1   женат / замужем   
4                0            53   среднее             1  гражданский брак   

   family_status_id gender employment_type  debt  monthly_income  \
0                 0      F       сотрудник     0          253875   
1                 0      F       сотрудник     0          112080   
2                 0      M       сотрудник     0          145885   
3                 0      M       сотрудник     0          267628   
4                 1      F       пенсионер     0          158616   

               credit_purpose  years_employed purpose_type  \
0               покупка жилья              23        жилье   
1     приобретение автомобиля              11   автомобиль   
2               покупка жилья              15        жилье   
3  дополнительное образование              11  образование   
4             сыграть свадьбу             932      свадьба   

     children_number_type  
0  семья с 1 или 2 детьми  
1  семья с 1 или 2 детьми  
2         семья без детей  
3       многодетная семья  
4         семья без детей

**Категоризация по ежемесячному доходу**

In [74]:
# обзор столбца `monthly_income`
data['monthly_income'].describe().astype(int)

count      21453
mean      165232
std        98213
min        20667
25%       107310
50%       143485
75%       195799
max      2265604
Name: monthly_income, dtype: int64

Можно категоризовать ежемесячный доход по квантилям:
1. 0% - 25%: низкий;
2. 25% - 75%: средний;
3. 75% - 100%: высокий.

In [75]:
# определение функции распределения ежемесячного дохода по заданным категориям
def monthly_income_categorizer(income):
    """Функция разбивает значения ежемесячного дохода на 3 категории: низкий, средний и высокий.
    """
    if income <= 100000:
        return 'низкий'
    elif 100000 < income <= 200000:
        return 'средний'
    else:
        return 'высокий'

In [76]:
# применение функции 
data['monthly_income_status'] = data['monthly_income'].apply(monthly_income_categorizer)

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Для выделения категорий по уровню дохода удобно использовать метод cut()/qcut(), [пример](https://dfedorov.spb.ru/pandas/Разделение%20данных%20в%20Pandas%20с%20помощью%20qcut%20и%20cut.html)
</div>


<div class="alert alert-info">

<b>Комментарий студента</b>

Хорошо, посмотрю
</div>

In [77]:
# проверка добавления столбца 'monthly_income_status'
data.head(5)

children_number  age_in_years education  education_id     family_status  \
0                1            42    высшее             0   женат / замужем   
1                1            36   среднее             1   женат / замужем   
2                0            33   среднее             1   женат / замужем   
3                3            32   среднее             1   женат / замужем   
4                0            53   среднее             1  гражданский брак   

   family_status_id gender employment_type  debt  monthly_income  \
0                 0      F       сотрудник     0          253875   
1                 0      F       сотрудник     0          112080   
2                 0      M       сотрудник     0          145885   
3                 0      M       сотрудник     0          267628   
4                 1      F       пенсионер     0          158616   

               credit_purpose  years_employed purpose_type  \
0               покупка жилья              23        жилье   
1     приобретение автомобиля              11   автомобиль   
2               покупка жилья              15        жилье   
3  дополнительное образование              11  образование   
4             сыграть свадьбу             932      свадьба   

     children_number_type monthly_income_status  
0  семья с 1 или 2 детьми               высокий  
1  семья с 1 или 2 детьми               средний  
2         семья без детей               средний  
3       многодетная семья               высокий  
4         семья без детей               средний

**Вывод**

Сгруппировав количество детей в семьях клиентов и ежемесячный доход, можно узнать вероятности того, будет ли задолженность по кредиту у клиентов, относящихся к определенным категориям.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Ты отлично справилась с добавлением категорий пользователей
</div>



## Шаг 3. Ответьте на вопросы

Перед тем как отвечать на вопросы о зависимости определенных параметров клиентов банка от потенциального наличия задолженности, необходимо определить функцию, которая будет вычислять эту вероятность.

In [78]:
# определение функции расчета вероятности получить задолженность по кредиту в зависимости от категории, к которой относится клиент
def category_debt(category):
    """Функция группирует столбец 'debt' по указанной категории,
    затем подсчитывает все факты наличия задолженности,
    делит на общее число клиентов (вычисляет среднее значение столбца 'debt)'.
    Далее преобразовывает данные в DataFrame и сортирует по возрастанию столбца 'debt'.
    """
    return ((data.groupby(category)['debt'].mean())*100).to_frame().sort_values(by='debt')

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [79]:
# применение функции category_debt к столбцу 'children_number_type'
category_debt('children_number_type')

debt
children_number_type            
семья без детей         7.544358
многодетная семья       8.157895
семья с 1 или 2 детьми  9.265359

**Вывод**

Да, зависимость есть. Более вероятно, что клиенты без детей погасят кредит в срок (вероятность получить долг-7.5%).

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все супер, но было бы неплохо посмотреть на размеры групп. Если группы маленькие, результаты анализа могут быть нерепрезентативны
</div>



<div class="alert alert-info">

<b>Комментарий студента</b>

Да, поняла. Число элементов в группах должно быть примерно одинаковым
</div>

<div class="alert alert-success">
    
<b> Комментарий от ревьюера 2</b>

Группы могут быть разного размера, главное, чтобы в них не было слишком мало элементов.


</div>


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [80]:
# применение функции category_debt к столбцу 'family_status'
category_debt('family_status')

debt
family_status                  
вдовец / вдова         6.569343
в разводе              7.112971
женат / замужем        7.545182
гражданский брак       9.349398
не женат / не замужем  9.750890

**Вывод**

Да, зависимость прослеживается. Более вероятно, что вдовцы/вдовы и разведенные клиенты погасят кредит в срок (6.6% и 7.1%), в отличие от людей, находящихся в неофициальных отношениях (9.3% и 9.8%).

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [81]:
# применение функции category_debt к столбцу 'monthly_income_status'
category_debt('monthly_income_status')

debt
monthly_income_status          
высокий                7.068115
низкий                 7.930108
средний                8.629654

**Вывод**

Из полученных данных следует, что большая вероятность (8.6%) получить задолженность по кредиту у людей со средним уровнем зарплаты.

- Как разные цели кредита влияют на его возврат в срок?

In [82]:
# применение функции category_debt к столбцу 'purpose_type'
category_debt('purpose_type')

debt
purpose_type           
ремонт         5.766063
жилье          7.085388
недвижимость   7.379249
строительство  7.667732
свадьба        8.003442
образование    9.220035
автомобиль     9.359034

**Вывод**

Менее вероятно, что клиент с целью "ремонт" не выплатит кредит вовремя (5.8%). В отличие от клиента, у которого взят кредит для оплаты целей из категории "автомобиль" (9.4%).

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Все верно
</div>



## Шаг 4. Общий вывод

Гипотеза о том, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок, ответ - влияет.

Клиенты с **самой низкой вероятностью** получить задолженность по кредиту:

- не имеющие детей;
- имеющие официально зарегестрированные отношениями (или вдовы/вдовцы/разведенные);
- с высоким ежемесячным доходом.

Клиенты с **самой высокой вероятностью** получить задолженность по кредиту:
    
- имеющие 1 или 2 ребенка;
- состоящие в неофициальном браке или без отношений;
- со средним ежемесячным доходом.

## Рекомендации банку

1. Записывать стаж работы в годах.
2. Добавить функцию валидации при внесении данных на сервер (базу данных) для избежания появления дубликатов.
3. Добавить проверку при вводе нереальных числовых значений на сайте/приложении банка: возраст, стаж, количество детей.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Хороший вывод
</div>



<div class="alert alert-success">
    
<b> Комментарий от ревьюера</b>
    
Спасибо за твою работу! Ты хорошо справилась с подготовкой и анализом данных. У тебя получились логичные комментарии и выводы.

Замечаний к проекту нет, но я верну работу на случай, если у тебя появились вопросы к моим комментариям. Если вопросов нет, можешь прислать проект без изменений, и я его приму.

</div>



<div class="alert alert-info">

<b>Комментарий студента</b>

Ярослав, спасибо за твои советы по улучшению кода :)
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Спасибо, что учла мои комментарии. Надеюсь, эта работа была полезна для тебя. Я принимаю проект и желаю удачи в прохождении курса!

</div> 

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.